components/datetime/src/pattern/item/ule.rs

```diff
@@ -0,0 +1,95 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::PatternItem;
+use crate::fields;
+use core::convert::TryFrom;
+use zerovec::ule::{AsULE, EqULE, ULE};
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct PatternItemULE(pub [u8; 3]);
+
+impl PatternItemULE {
+    pub fn is_literal_from_u8(byte: u8) -> bool {
+        byte & 0b1000_0000 != 0
+    }
+
+    pub fn clear_type_in_u8(byte: u8) -> u8 {
+        byte ^ 0b1000_0000
+    }
+
+    pub fn set_literal_in_u8(byte: u8) -> u8 {
+        byte | 0b1000_0000
+    }
+
+    pub fn bytes_in_range(value: (&u8, &u8, &u8)) -> bool {
+        match Self::is_literal_from_u8(*value.0) {
+            false => fields::Field::bytes_in_range(value.1, value.2),
+            true => {
+                let u = u32::from_le_bytes([
+                    *value.2,
+                    *value.1,
+                    Self::clear_type_in_u8(*value.0),
+                    0x00,
+                ]);
+                char::try_from(u).is_ok()
+            }
+        }
+    }
+}
+
+unsafe impl ULE for PatternItemULE {
+    type Error = &'static str;
+
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error> {
+        let mut chunks = bytes.chunks_exact(3);
+
+        if !chunks.all(|c| Self::bytes_in_range((&c[0], &c[1], &c[2])))
+            || !chunks.remainder().is_empty()
+        {
+            return Err("Invalid byte sequence");
+        }
+        Ok(())
+    }
+}
+
+impl AsULE for PatternItem {
+    type ULE = PatternItemULE;
+
+    #[inline]
+    fn as_unaligned(&self) -> Self::ULE {
+        match self {
+            Self::Field(field) => PatternItemULE([0x00, field.symbol.idx(), field.length as u8]),
+            Self::Literal(ch) => {
+                let u = *ch as u32;
+                let bytes = u.to_be_bytes();
+                PatternItemULE([
+                    PatternItemULE::set_literal_in_u8(bytes[1]),
+                    bytes[2],
+                    bytes[3],
+                ])
+            }
+        }
+    }
+
+    #[inline]
+    fn from_unaligned(unaligned: &Self::ULE) -> Self {
+        let value = unaligned.0;
+        match PatternItemULE::is_literal_from_u8(value[0]) {
+            false => {
+                let symbol = fields::FieldSymbol::from_idx(value[1]).unwrap();
```

Comment:
Discussion: should we have a `from_idx_unchecked` function that we call here, at least in production mode, rather than unwrapping?

Reply:
Fair enough.  I was asking more as a style question rather than a metrics question, but having metrics could help (and potentially for code size impact).